src/aspire/orientation/commonline_sync.py

```diff
@@ -1,31 +1,443 @@
+            # condition below
+            #       1+2*c1*c2*c3-(c1^2+c2^2+c3^2) > 1.0e-5.
+            # This ensures that the smallest singular value (which is actually
+            # controlled by the determinant of C) is big enough, so the matrix is far
+            # from singular. This condition is equivalent to computing the singular
+            # values of C, followed by checking that the smallest one is big enough.
+            cond = 1 + 2 * c1 * c2 * c3 - (
+                    np.square(c1) + np.square(c2) + np.square(c3))
+
+            good_idx = np.where(cond > 1e-5)[0]
+            bad_idx = np.where(cond <= 1e-5)[0]
+
+            cos_phi2 = (c3[good_idx] - c1[good_idx] *
+                        c2[good_idx]) / (np.sin(theta1[good_idx])
+                                         * np.sin(theta2[good_idx]))
+            check_idx = np.where(np.abs(cos_phi2) > 1)[0]
+            if np.any(np.abs(cos_phi2) - 1 > 1e-12):
+                logger.warning(f'Globally Consistent Angular Reconstruction(GCAR) exists'
+                               ' numerical problem: abs(cos_phi2) >1, with the '
+                               ' difference of {np.abs(cos_phi2)-1}.')
+            elif len(check_idx) == 0:
+                cos_phi2[check_idx] = np.sign(cos_phi2[check_idx])
+
+            phis[:idx + len(good_idx), 0] = cos_phi2
+            phis[:idx + len(good_idx), 1] = k_list[good_idx]
+            idx += len(good_idx)
+
+            rejected[: rej_idx + len(bad_idx)] = k_list[bad_idx]
+            rej_idx += len(bad_idx)
+
+        phis = phis[:idx]
+        rejected = rejected[:rej_idx]
+
+        good_k = []
+        peakh = -1
+        alpha = []
+
+        if idx > 0:
+            # Compute the histogram of the angles between images i and j.
+            angles = np.arccos(phis[:, 0]) * 180 / np.pi
+            # Angles that are up to 10 degrees apart are considered
+            # similar. This sigma ensures that the width of the density
+            # estimation kernel is roughly 10 degrees. For 15 degrees, the
+            # value of the kernel is negligible.
+            sigma = 3.0
+
+            tmp = np.add.outer(np.square(angles), np.square(x))
+            h = np.sum(np.exp((2 * np.multiply.outer(angles, x)
+                               - tmp) / (2 * sigma ** 2)), 0)
+
+            # We assume that at the location of the peak we get the true angle
+            # between images i and j. Find all third images k, that induce an
+            # angle between i and j that is at most 10 off the true angle.
+            # Even for debugging, don't put a value that is smaller than two
+            # tics, since the peak might move a little bit due to wrong k images
+            # that accidentally fall near the peak.
+            peak_idx = h.argmax()
+            peakh = h[peak_idx]
+            idx = np.where(np.abs(angles - x[peak_idx]) < 360 / ntics)[0]
```

Comment:
It looks that it needs to find a list of indices. Removing doesn't work. I will keep it.

Reply:
Really. NumPy allows you to index into an array using an array of booleans. So you're saying that
```
idx = (np.abs(angles - x[peak_idx]) < 360 / ntics)
good_k = inds[idx]
```
fails? What error do you get?

Reply:
IndexError: boolean index did not match indexed array along dimension 0; dimension is 64 but corresponding boolean dimension is 20.   Here we only need to keep the good indices instead of a mask with the full length.

Reply:
That sounds like a bug, then. We should have the same length for `idx`, `angles`, and `phis` (aka `cos_phi2`), while the length of `inds` should be the same as `good_idx`. In `_get_cos_phis`, both `cos_phi2` and `good_idx` have the same length, so what's missing?

Reply:
Done. the `[0]` in the original one should be removed too.